vital/tests/test_mesh.cxx

```diff
@@ -0,0 +1,236 @@
+
+  for(unsigned int i=0; i<first_list.size(); i++)
+  {
+    for(unsigned int j=0; j<first_list[i].size(); j++)
+    {
+      EXPECT_EQ( first_list[i][j], first_faces[i][j] );
+    }
+  }
+
+  unsigned int offset = first_list.size();
+  for(unsigned int i=0; i<second_list.size(); i++)
+  {
+    for(unsigned int j=0; j<second_list[i].size(); j++)
+    {
+      EXPECT_EQ( second_list[i][j] + shift, first_faces[i + offset][j] );
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+TEST(mesh, half_edges)
+{
+  kwiver::vital::mesh_sptr cube_mesh = kwiver::testing::cube_mesh( 1.0 );
+
+  std::vector<std::vector<unsigned int>> face_list = { {0, 1, 3, 2},
+                                                       {4, 6, 7, 5},
+                                                       {5, 7, 3, 1},
+                                                       {6, 4, 0, 2},
+                                                       {7, 6, 2, 3},
+                                                       {1, 0, 4, 5} };
+  unsigned int list_size = 0;
+  for(unsigned int i=0; i<face_list.size(); i++)
+  {
+    list_size+=face_list[i].size();
+  }
+
+  kwiver::vital::mesh_half_edge_set constructed_edges( face_list );
+  EXPECT_FALSE( cube_mesh->has_half_edges() );
+  cube_mesh->build_edge_graph();
+  EXPECT_TRUE( cube_mesh->has_half_edges() );
+  kwiver::vital::mesh_half_edge_set copy_edges = cube_mesh->half_edges();
+
+  EXPECT_EQ( constructed_edges.num_verts(), copy_edges.num_verts() );
+  EXPECT_EQ( constructed_edges.num_faces(), copy_edges.num_faces() );
+
+  EXPECT_EQ( copy_edges.size(), list_size );
+  EXPECT_EQ( cube_mesh->num_faces(), face_list.size() );
+}
+
+// ----------------------------------------------------------------------------
+TEST(mesh, copy_constructor)
+{
+  kwiver::vital::mesh_sptr original = kwiver::testing::cube_mesh( 1.0 );
+  kwiver::vital::mesh_sptr copy =
+    std::make_shared<kwiver::vital::mesh>( *original );
+
+  compare_meshes( original, copy );
+
+  EXPECT_TRUE( original->is_init() );
+  EXPECT_TRUE( copy->is_init() );
+  *original = kwiver::vital::mesh();
+  EXPECT_FALSE( original->is_init() );
+  EXPECT_TRUE( copy->is_init() );
+}
+
+// ----------------------------------------------------------------------------
+TEST(mesh, assignment_operator)
+{
+  kwiver::vital::mesh_sptr original = kwiver::testing::cube_mesh( 1.0 );
+  kwiver::vital::mesh_sptr copy =
+    std::make_shared<kwiver::vital::mesh>();
+  *copy = kwiver::vital::mesh( *original );
+
+  compare_meshes( original, copy );
+
+  EXPECT_TRUE( original->is_init() );
+  EXPECT_TRUE( copy->is_init() );
+  *original = kwiver::vital::mesh();
+  EXPECT_FALSE( original->is_init() );
+  EXPECT_TRUE( copy->is_init() );
+}
+
+// ----------------------------------------------------------------------------
+TEST(mesh, compute_vertex_normals)
+{
+  kwiver::vital::mesh_sptr grid_mesh = kwiver::testing::grid_mesh( 1, 1 );
```

Comment:
Maybe make this grid mesh at least (2,3).  A (1,1) mesh is just a single face.  There might be cases not covered for such a basic case.

Actually cube might be better here since different faces will have different normals.

Maybe even better is a mesh that appends a cube to a grid mesh.  The cube is a closed shape with no boundary edges which may have different special cases from the grid mesh which does have a boundary and all faces have the same normal.  It might be best to cover both.